# 第五题（选做）：完成dota2比赛结果预测任务

实验内容：
1. 使用你学过的算法完成dota2比赛结果预测
2. 对比它们的精度、查准率、查全率、F1值

## 结果写入到markdown的表格中！

In [1]:
import numpy as np
data = np.loadtxt('data/dota2Dataset/dota2Train.csv', delimiter=',')
print(data.shape)

(92650, 117)


In [2]:
# 数据切分
# 训练集测试集三七分
size = len(data)
data_y = data[:,0]
data_x = data[:,1:]

trainX = data_x[:int(size*0.7),:]
trainY = data_y[:int(size*0.7)]

testX = data_x[int(size*0.7):,:]
testY = data_y[int(size*0.7):]

print(trainX.shape)
print(trainY.shape)
print(testX.shape)
print(testY.shape)

(64854, 116)
(64854,)
(27796, 116)
(27796,)


In [3]:
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score

In [4]:
clf = SVC(kernel = 'rbf', C = 1, random_state = 32);

clf.fit(trainX,trainY);

prediction = clf.predict(testX);

print(accuracy_score(testY,prediction))
print(precision_score(testY,prediction))
print(recall_score(testY,prediction))
print(f1_score(testY,prediction))

0.593862426248381
0.6019190404797601
0.6833469498910676
0.6400535663042439


In [5]:
clf = SVC(kernel = 'rbf', C = 0.1, random_state = 32)

clf.fit(trainX,trainY)

prediction = clf.predict(testX)

print(accuracy_score(testY,prediction))
print(precision_score(testY,prediction))
print(recall_score(testY,prediction))
print(f1_score(testY,prediction))

0.5686789466110231
0.5576143107202323
0.889229302832244
0.6854189079268452


In [6]:
clf = LinearSVC(C = 1, random_state = 32)

clf.fit(trainX,trainY)

prediction = clf.predict(testX)

print(accuracy_score(testY,prediction))
print(precision_score(testY,prediction))
print(recall_score(testY,prediction))
print(f1_score(testY,prediction))

0.5294646711757087
0.5289962155343305
0.9992510893246187
0.691772912591615


In [7]:
clf = LinearSVC(C = 0.1, random_state = 32)

clf.fit(trainX,trainY)

prediction = clf.predict(testX)

print(accuracy_score(testY,prediction))
print(precision_score(testY,prediction))
print(recall_score(testY,prediction))
print(f1_score(testY,prediction))

0.532306806734782
0.5305597798537186
0.9976171023965141
0.6927149813265259


In [8]:
# YOUR CODE HERE
clf = SVC(kernel = 'rbf', C = 0.1, random_state = 32)

clf.fit(trainX,trainY)

prediction = clf.predict(testX)

print(accuracy_score(testY,prediction))
print(precision_score(testY,prediction))
print(recall_score(testY,prediction))
print(f1_score(testY,prediction))

0.5686789466110231
0.5576143107202323
0.889229302832244
0.6854189079268452


In [9]:
clf = SVC(kernel = 'rbf', C = 1, random_state = 32)

clf.fit(trainX,trainY)

prediction = clf.predict(testX)

print(accuracy_score(testY,prediction))
print(precision_score(testY,prediction))
print(recall_score(testY,prediction))
print(f1_score(testY,prediction))

0.593862426248381
0.6019190404797601
0.6833469498910676
0.6400535663042439


核函数 | C | 精度 | 查准率 | 查全率 | F1
- | - | - | - | - | -
rbf | 0.1 | 0.5686789466110231 | 0.5576143107202323 | 0.889229302832244 | 0.6854189079268452
rbf | 1 | 0.593862426248381 | 0.6019190404797601 | 0.6833469498910676 | 0.6400535663042439
linear | 0.1 | 0.532306806734782 | 0.5305597798537186 | 0.9976171023965141 | 0.6927149813265259
linear | 1 | 0.5294646711757087 | 0.5289962155343305 | 0.9992510893246187 | 0.691772912591615
sigmoid | 0.1 | 0.5686789466110231 | 0.5576143107202323 | 0.889229302832244 | 0.6854189079268452
sigmoid | 1 | 0.593862426248381 | 0.6019190404797601 | 0.6833469498910676 | 0.6400535663042439